rust-runtime/smithy-json/src/deserialize/token.rs

```diff
@@ -0,0 +1,362 @@
+            Some(Token::StartObject { .. }) | Some(Token::StartArray { .. }) => {
+                skip_inner(true, tokens)?;
+                if !inside_obj_or_array {
+                    break;
+                }
+            }
+            Some(Token::EndObject { .. }) | Some(Token::EndArray { .. }) => {
+                debug_assert!(inside_obj_or_array);
+                break;
+            }
+            Some(Token::ValueNull { .. })
+            | Some(Token::ValueBool { .. })
+            | Some(Token::ValueNumber { .. })
+            | Some(Token::ValueString { .. }) => {
+                if !inside_obj_or_array {
+                    break;
+                }
+            }
+            Some(Token::ObjectKey { .. }) => {}
+            _ => return Err(Error::custom("expected value")),
+        }
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+pub mod test {
+    use super::*;
+    use crate::deserialize::error::ErrorReason::UnexpectedToken;
+    use crate::deserialize::json_token_iter;
+
+    pub fn start_array<'a>(offset: usize) -> Option<Result<Token<'a>, Error>> {
+        Some(Ok(Token::StartArray {
+            offset: TokenOffset(offset),
+        }))
+    }
+
+    pub fn end_array<'a>(offset: usize) -> Option<Result<Token<'a>, Error>> {
+        Some(Ok(Token::EndArray {
+            offset: TokenOffset(offset),
+        }))
+    }
+
+    pub fn start_object<'a>(offset: usize) -> Option<Result<Token<'a>, Error>> {
+        Some(Ok(Token::StartObject {
+            offset: TokenOffset(offset),
+        }))
+    }
+
+    pub fn end_object<'a>(offset: usize) -> Option<Result<Token<'a>, Error>> {
+        Some(Ok(Token::EndObject {
+            offset: TokenOffset(offset),
+        }))
+    }
+
+    pub fn object_key<'a>(offset: usize, key: &'a str) -> Option<Result<Token<'a>, Error>> {
+        Some(Ok(Token::ObjectKey {
+            offset: TokenOffset(offset),
+            key: EscapedStr::new(key),
+        }))
+    }
+
+    pub fn value_bool<'a>(offset: usize, boolean: bool) -> Option<Result<Token<'a>, Error>> {
+        Some(Ok(Token::ValueBool {
+            offset: TokenOffset(offset),
+            value: boolean,
+        }))
+    }
+
+    pub fn value_number<'a>(offset: usize, number: Number) -> Option<Result<Token<'a>, Error>> {
+        Some(Ok(Token::ValueNumber {
+            offset: TokenOffset(offset),
+            value: number,
+        }))
+    }
+
+    pub fn value_null<'a>(offset: usize) -> Option<Result<Token<'a>, Error>> {
+        Some(Ok(Token::ValueNull {
+            offset: TokenOffset(offset),
+        }))
+    }
+
+    pub fn value_string<'a>(offset: usize, string: &'a str) -> Option<Result<Token<'a>, Error>> {
+        Some(Ok(Token::ValueString {
+            offset: TokenOffset(offset),
+            value: EscapedStr::new(string),
+        }))
+    }
+
+    #[test]
+    fn skip_simple_value() {
+        let mut tokens = json_token_iter(b"null true");
+        skip_value(&mut tokens).unwrap();
```

Comment:
so it seems like json_token_iter accepts invalid JSON? not necessarily a problem, just worth documenting what is accepted and what is not

Reply:
I don't think that's necessarily invalid, but rather, multiple values. I could make it only read one value and then fail if there's more, but that's effectively what the codegen doing with it.